# STEP1 Mounting Drive





In [1]:
from google.colab import drive    #You can access your drive files by mounting the drive
drive.mount('/content/drive')

Mounted at /content/drive


# STEP2 Checking Hardware Configurations

In [2]:
!nvidia-smi     # checking hardware

Wed Jan  6 07:30:25 2021       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.27.04    Driver Version: 418.67       CUDA Version: 10.1     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla T4            Off  | 00000000:00:04.0 Off |                    0 |
| N/A   46C    P8    10W /  70W |      0MiB / 15079MiB |      0%      Default |
|                               |                      |                 ERR! |
+-------------------------------+----------------------+----------------------+
                                                                               
+-------

In [3]:
!nvcc -V

nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Sun_Jul_28_19:07:16_PDT_2019
Cuda compilation tools, release 10.1, V10.1.243


In [4]:

!python -V; pip -V      #checking versions of python and pip

Python 3.6.9
pip 19.3.1 from /usr/local/lib/python3.6/dist-packages/pip (python 3.6)


# STEP3 Installation of cudf

In [5]:

!git clone https://github.com/rapidsai/rapidsai-csp-utils.git
!bash rapidsai-csp-utils/colab/rapids-colab.sh

import sys, os

dist_package_index = sys.path.index('/usr/local/lib/python3.6/dist-packages')
sys.path = sys.path[:dist_package_index] + ['/usr/local/lib/python3.6/site-packages'] + sys.path[dist_package_index:]
sys.path
exec(open('rapidsai-csp-utils/colab/update_modules.py').read(), globals())

Cloning into 'rapidsai-csp-utils'...
remote: Enumerating objects: 14, done.
remote: Counting objects: 100% (14/14), done.
remote: Compressing objects: 100% (14/14), done.
remote: Total 185 (delta 5), reused 3 (delta 0), pack-reused 171
Receiving objects: 100% (185/185), 57.32 KiB | 11.46 MiB/s, done.
Resolving deltas: 100% (67/67), done.
PLEASE READ
********************************************************************************************************
Changes:
1. IMPORTANT CHANGES: RAPIDS on Colab will be pegged to 0.14 Stable until further notice.
2. Default stable version is now 0.14.  Nightly will redirect to 0.14.
3. You can now declare your RAPIDSAI version as a CLI option and skip the user prompts (ex: '0.14' or '0.15', between 0.13 to 0.14, without the quotes): 
        "!bash rapidsai-csp-utils/colab/rapids-colab.sh <version/label>"
        Examples: '!bash rapidsai-csp-utils/colab/rapids-colab.sh 0.14', or '!bash rapidsai-csp-utils/colab/rapids-colab.sh stable', or '!bash rap

#STEP4  IMP Libraries

In [6]:
from scipy import stats
import time
import numpy as np
import pandas as pd


#STEP5  Working Link

In [7]:
link='/content/drive/MyDrive/5m.csv'

#STEP6 Working with CUDF

In [9]:
import cudf              #reading our file with cudf
start=time.time()
df_cudf=cudf.read_csv(link)
end=time.time()
print("Time taken by cudf to load dataframe is",end='   ')
print(end-start,end=' ')
print("seconds")

Time taken by cudf to load dataframe is   1.6767680644989014 seconds


In [10]:
df_cudf.head()   #Visualising the cudf dataframe

Region  Country  ...  Total Cost Total Profit
0         Australia and Oceania    Palau  ...  1260428.96    303126.25
1                        Europe   Poland  ...   296918.60    146264.40
2                 North America   Canada  ...    12062.33      9124.77
3                        Europe  Belarus  ...   137780.16     77967.96
4  Middle East and North Africa     Oman  ...   822931.97    622521.93

[5 rows x 14 columns]

In [11]:
type(df_cudf)   #Type of the dataframe

cudf.core.dataframe.DataFrame

# STEP7 Finding Mean

In [12]:
start=time.time()                    
Total_Profit_mean = df_cudf['Total Profit'].mean()
Units_Sold_mean = df_cudf['Units Sold'].mean()
Unit_Price_mean = df_cudf['Unit Price'].mean()
Unit_Cost_mean = df_cudf['Unit Cost'].mean()
Total_Revenue_mean = df_cudf['Total Revenue'].mean()
Total_Cost_mean = df_cudf['Total Cost'].mean()
end=time.time()
print("Time taken by cudf to Calculate Mean is",end='   ')
print(end-start,end=' ')
print("seconds")
print("Total_Profit_mean is %d"%Total_Profit_mean)
print("units_sold_mean is %d"%Units_Sold_mean)
print("Unit_Price_mean is %d"%Unit_Price_mean)
print("Unit_Cost_mean is %d"%Unit_Cost_mean)
print("Total_Revenue_mean is %d"%Total_Revenue_mean)
print("Total_Cost_mean is %d"%Total_Cost_mean)

Time taken by cudf to Calculate Mean is   0.0063359737396240234 seconds
Total_Profit_mean is 392679
units_sold_mean is 4999
Unit_Price_mean is 266
Unit_Cost_mean is 187
Total_Revenue_mean is 1331058
Total_Cost_mean is 938378


ENTITIES | MEAN
--- | ---
 Total_Profit_mean |  392679
Units_Sold_mean | 4999
Unit_Price_mean | 266
Unit_Cost_mean | 187
Total_Revenue_mean | 1331058
Total_Cost_mean | 938378

#STEP7 Median

In [13]:
start=time.time()                    
Total_Profit_median = df_cudf['Total Profit'].median()
Units_Sold_median = df_cudf['Units Sold'].median()
Unit_Price_median = df_cudf['Unit Price'].median()
Unit_Cost_median = df_cudf['Unit Cost'].median()
Total_Revenue_median = df_cudf['Total Revenue'].median()
Total_Cost_median = df_cudf['Total Cost'].median()
end=time.time()
print(end-start)
print("Time taken by cudf to Calculate Median is",end='   ')
print(end-start,end=' ')
print("seconds")
print("Total_Profit_median is %d"%Total_Profit_median)
print("units_sold_median is %d"%Units_Sold_median)
print("Unit_Price_median is %d"%Unit_Price_median)
print("Unit_Cost_median is %d"%Unit_Cost_median)
print("Total_Revenue_median is %d"%Total_Revenue_median)
print("Total_Cost_median is %d"%Total_Cost_median)

0.06610488891601562
Time taken by cudf to Calculate Median is   0.06610488891601562 seconds
Total_Profit_median is 281655
units_sold_median is 4999
Unit_Price_median is 205
Unit_Cost_median is 117
Total_Revenue_median is 785979
Total_Cost_median is 467712


ENTITIES | MEDIAN
--- | ---
 Total_Profit_median |  281655
Units_Sold_median | 4999
Unit_Price_median | 205
Unit_Cost_median | 117
Total_Revenue_median | 785979
Total_Cost_median | 467712

# STEP8 Artihmetic between rows

In [14]:
start=time.time()
df_cudf['Checking Total profit']=df_cudf['Units Sold']*df_cudf['Unit Price'] - df_cudf['Total Cost']
end=time.time()
print("Time taken by cudf to Calculate Arithmetic mean between two rows is ",end='   ')
print(end-start,end=' ')
print("seconds")

Time taken by cudf to Calculate Arithmetic mean between two rows is    0.5063333511352539 seconds


In [15]:
df_cudf.head()

Region  Country  ... Total Profit Checking Total profit
0         Australia and Oceania    Palau  ...    303126.25             303126.25
1                        Europe   Poland  ...    146264.40             146264.40
2                 North America   Canada  ...      9124.77               9124.77
3                        Europe  Belarus  ...     77967.96              77967.96
4  Middle East and North Africa     Oman  ...    622521.93             622521.93

[5 rows x 15 columns]

In [16]:
df_cudf.drop('Checking Total profit', axis=1,inplace=True)

Region  ... Total Profit
0               Australia and Oceania  ...    303126.25
1                              Europe  ...    146264.40
2                       North America  ...      9124.77
3                              Europe  ...     77967.96
4        Middle East and North Africa  ...    622521.93
...                               ...  ...          ...
4999995  Middle East and North Africa  ...    809425.32
4999996                        Europe  ...    179722.40
4999997         Australia and Oceania  ...    324531.36
4999998         Australia and Oceania  ...     49031.34
4999999                        Europe  ...    546536.25

[5000000 rows x 14 columns]

# STEP9 Checking dataframe types

In [17]:
type(df_cudf)

cudf.core.dataframe.DataFrame

In [18]:
dfd=df_cudf.iloc[:,8]

In [19]:
type(dfd)

cudf.core.series.Series

# STEP10 Time while calculating using numpy function

In [20]:
import numpy as np

In [21]:
start=time.time()
d=np.mean(dfd.values)
end=time.time()
print(end-start)

0.21920132637023926


In [22]:
type(d)

cupy.core.core.ndarray

In [23]:
type(dfd.values)

cupy.core.core.ndarray

In [24]:
dfd.values.shape

(5000000,)

# STEP 11 Working With DaskCUDF Installation

In [25]:
pip install dask[complete]

# STEP 12 Importing the DataFrame

In [26]:
import dask_cudf
start=time.time()
df_daskcudf=dask_cudf.read_csv(link)
end=time.time()
print("Time taken by Dask Cudf to load dataframe is",end='   ')
print(end-start,end=' ')
print("seconds")

Time taken by Dask Cudf to load dataframe is   0.06547737121582031 seconds


In [27]:
df_daskcudf.head()

Region  Country  ...  Total Cost Total Profit
0         Australia and Oceania    Palau  ...  1260428.96    303126.25
1                        Europe   Poland  ...   296918.60    146264.40
2                 North America   Canada  ...    12062.33      9124.77
3                        Europe  Belarus  ...   137780.16     77967.96
4  Middle East and North Africa     Oman  ...   822931.97    622521.93

[5 rows x 14 columns]

In [28]:
type(df_daskcudf)

dask_cudf.core.DataFrame

# STEP12 Finding Mean

In [29]:
start=time.time()                    
Total_Profit_mean = df_daskcudf['Total Profit'].compute().mean()
Units_Sold_mean = df_daskcudf['Units Sold'].compute().mean()
Unit_Price_mean = df_daskcudf['Unit Price'].compute().mean()
Unit_Cost_mean = df_daskcudf['Unit Cost'].compute().mean()
Total_Revenue_mean = df_daskcudf['Total Revenue'].compute().mean()
Total_Cost_mean = df_daskcudf['Total Cost'].compute().mean()
end=time.time()
print("Time taken by Dask Cudf to Calculate Mean is",end='   ')
print(end-start,end=' ')
print("seconds")
print("Total_Profit_mean is %d"%Total_Profit_mean)
print("units_sold_mean is %d"%Units_Sold_mean)
print("Unit_Price_mean is %d"%Unit_Price_mean)
print("Unit_Cost_mean is %d"%Unit_Cost_mean)
print("Total_Revenue_mean is %d"%Total_Revenue_mean)
print("Total_Cost_mean is %d"%Total_Cost_mean)

Time taken by Dask Cudf to Calculate Mean is   10.021194458007812 seconds
Total_Profit_mean is 392679
units_sold_mean is 4999
Unit_Price_mean is 266
Unit_Cost_mean is 187
Total_Revenue_mean is 1331058
Total_Cost_mean is 938378


ENTITIES | MEAN
--- | ---
 Total_Profit_mean |  392679
Units_Sold_mean | 4999
Unit_Price_mean | 266
Unit_Cost_mean | 187
Total_Revenue_mean | 1331058
Total_Cost_mean | 938378

# Step13 Finding Median

In [30]:
start=time.time()                    
Total_Profit_median = df_daskcudf['Total Profit'].compute().median()
Units_Sold_median = df_daskcudf['Units Sold'].compute().median()
Unit_Price_median = df_daskcudf['Unit Price'].compute().median()
Unit_Cost_median = df_daskcudf['Unit Cost'].compute().median()
Total_Revenue_median = df_daskcudf['Total Revenue'].compute().median()
Total_Cost_median = df_daskcudf['Total Cost'].compute().median()
end=time.time()
print(end-start)
print("Time taken by Daskcudf to Calculate Median is",end='   ')
print(end-start,end=' ')
print("seconds")
print("Total_Profit_median is %d"%Total_Profit_median)
print("units_sold_median is %d"%Units_Sold_median)
print("Unit_Price_median is %d"%Unit_Price_median)
print("Unit_Cost_median is %d"%Unit_Cost_median)
print("Total_Revenue_median is %d"%Total_Revenue_median)
print("Total_Cost_median is %d"%Total_Cost_median)

10.08503007888794
Time taken by Daskcudf to Calculate Median is   10.08503007888794 seconds
Total_Profit_median is 281655
units_sold_median is 4999
Unit_Price_median is 205
Unit_Cost_median is 117
Total_Revenue_median is 785979
Total_Cost_median is 467712


ENTITIES | MEDIAN
--- | ---
 Total_Profit_median |  281655
Units_Sold_median | 4999
Unit_Price_median | 205
Unit_Cost_median | 117
Total_Revenue_median | 785979
Total_Cost_median | 467712

# STEP14 Artihmetic between rows

In [31]:
start=time.time()
df_daskcudf['Chechking Total profit']= df_daskcudf['Units Sold']* df_daskcudf['Unit Price'] -  df_daskcudf['Total Cost']
end=time.time()
print("Time taken by DaskDataframe to Calculate Arithmetic mean between two rows is ",end='   ')
print(end-start,end=' ')
print("seconds")

Time taken by DaskDataframe to Calculate Arithmetic mean between two rows is    0.023401260375976562 seconds


In [32]:
 df_daskcudf.compute().drop('Chechking Total profit', axis=1,inplace=True)

Region  ... Total Profit
0              Australia and Oceania  ...    303126.25
1                             Europe  ...    146264.40
2                      North America  ...      9124.77
3                             Europe  ...     77967.96
4       Middle East and North Africa  ...    622521.93
...                              ...  ...          ...
698093  Middle East and North Africa  ...    809425.32
698094                        Europe  ...    179722.40
698095         Australia and Oceania  ...    324531.36
698096         Australia and Oceania  ...     49031.34
698097                        Europe  ...    546536.25

[5000000 rows x 14 columns]

# STEP15 Checking dataframe types

In [33]:
dfd=df_daskcudf.iloc[:,8]

In [34]:
type(dfd)

dask_cudf.core.Series

# STEP 16 Time calculation using Numpy function

In [35]:
import numpy as np
start=time.time()
d=np.mean(dfd.values)
end=time.time()
print(end-start)

0.009670257568359375


In [36]:
type( df_daskcudf)

dask_cudf.core.DataFrame

# COMPARISON IN PERFORMANCE 

ENTITIES | CUDF | DASKCUDF
--- | --- | ---
Time in loading the dataset frame |  1.8919148445129395 | 0.0536806583404541
Time in calculating mean| 0.0046367645263671875 | 10.659050226211548
Time in calculating median | 0.04202437400817871 |  10.206039667129517
Time in calculating arithmetic mean | 0.5742025375366211 |  0.023900270462036133
Type of dataframe loaded | cudf.core.dataframe.DataFrame | dask_cudf.core.DataFrame

# CONCLUSION

DASKCUDF IS BETTER THAN TO RAPIDSCUDF WHILE READING WHILE LOADING THE DATASET